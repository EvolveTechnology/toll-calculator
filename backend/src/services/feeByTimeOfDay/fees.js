const asMinutes = (h = 0, m = 0) => h * 60 + m;
const minutesRange = (...from) => (...to) => [from, to].map(hour => asMinutes(...hour));

export default [
  { range: minutesRange(6)(6, 29), fee: 8 },
  { range: minutesRange(6, 30)(6, 59), fee: 13 },
  { range: minutesRange(7)(7, 59), fee: 18 },
  { range: minutesRange(8)(8, 29), fee: 13 },
  { range: minutesRange(8, 30)(14, 59), fee: 8 },
  { range: minutesRange(15)(15, 29), fee: 13 },
  { range: minutesRange(15, 30)(16, 29), fee: 18 },
  { range: minutesRange(17)(17, 59), fee: 13 },
  { range: minutesRange(18)(18, 29), fee: 8 },
];
