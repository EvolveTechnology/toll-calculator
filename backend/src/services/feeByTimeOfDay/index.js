// Original Java specs
export default (date) => {
  const hour = date.getUTCHours();
  const minute = date.getUTCMinutes();

  if (hour === 6 && minute >= 0 && minute <= 29) return 8;
  if (hour === 6 && minute >= 30 && minute <= 59) return 13;
  if (hour === 7 && minute >= 0 && minute <= 59) return 18;
  if (hour === 8 && minute >= 0 && minute <= 29) return 13;
  if (hour >= 8 && hour <= 14 && minute >= 30 && minute <= 59) return 8;
  if (hour === 15 && minute >= 0 && minute <= 29) return 13;
  if ((hour === 15 && minute >= 0) || (hour === 16 && minute <= 59)) return 18;
  if (hour === 17 && minute >= 0 && minute <= 59) return 13;
  if (hour === 18 && minute >= 0 && minute <= 29) return 8;
  return 0;
};
