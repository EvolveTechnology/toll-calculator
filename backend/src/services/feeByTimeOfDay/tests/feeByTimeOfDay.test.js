import feeByTimeOfDay from '..';

describe('feeByTimeOfDay', () => {
  const baseDate = new Date(Date.UTC(2018, 0, 1));
  const baseDateUnix = baseDate.getTime();

  // define 30 minutes
  const halfHour = 30 * 60 * 1000;
  // generate timestamps separated by 30 minutes
  const aWholeDay = Array.from({ length: 48 }, (_, i) => new Date(baseDateUnix + i * halfHour));

  // use this to store the initially proposed pricing scheme
  it('calculates tool fees for different hours', () => {
    expect(aWholeDay.map(hour => ({ fee: feeByTimeOfDay(hour), hour }))).toMatchSnapshot();
  });
});
