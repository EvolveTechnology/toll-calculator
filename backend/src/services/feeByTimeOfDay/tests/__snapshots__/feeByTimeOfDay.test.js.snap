// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feeByTimeOfDay calculates tool fees for different hours 1`] = `
Array [
  Object {
    "fee": 0,
    "hour": 2018-01-01T00:00:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T00:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T01:00:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T01:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T02:00:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T02:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T03:00:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T03:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T04:00:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T04:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T05:00:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T05:30:00.000Z,
  },
  Object {
    "fee": 8,
    "hour": 2018-01-01T06:00:00.000Z,
  },
  Object {
    "fee": 13,
    "hour": 2018-01-01T06:30:00.000Z,
  },
  Object {
    "fee": 18,
    "hour": 2018-01-01T07:00:00.000Z,
  },
  Object {
    "fee": 18,
    "hour": 2018-01-01T07:30:00.000Z,
  },
  Object {
    "fee": 13,
    "hour": 2018-01-01T08:00:00.000Z,
  },
  Object {
    "fee": 8,
    "hour": 2018-01-01T08:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T09:00:00.000Z,
  },
  Object {
    "fee": 8,
    "hour": 2018-01-01T09:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T10:00:00.000Z,
  },
  Object {
    "fee": 8,
    "hour": 2018-01-01T10:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T11:00:00.000Z,
  },
  Object {
    "fee": 8,
    "hour": 2018-01-01T11:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T12:00:00.000Z,
  },
  Object {
    "fee": 8,
    "hour": 2018-01-01T12:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T13:00:00.000Z,
  },
  Object {
    "fee": 8,
    "hour": 2018-01-01T13:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T14:00:00.000Z,
  },
  Object {
    "fee": 8,
    "hour": 2018-01-01T14:30:00.000Z,
  },
  Object {
    "fee": 13,
    "hour": 2018-01-01T15:00:00.000Z,
  },
  Object {
    "fee": 18,
    "hour": 2018-01-01T15:30:00.000Z,
  },
  Object {
    "fee": 18,
    "hour": 2018-01-01T16:00:00.000Z,
  },
  Object {
    "fee": 18,
    "hour": 2018-01-01T16:30:00.000Z,
  },
  Object {
    "fee": 13,
    "hour": 2018-01-01T17:00:00.000Z,
  },
  Object {
    "fee": 13,
    "hour": 2018-01-01T17:30:00.000Z,
  },
  Object {
    "fee": 8,
    "hour": 2018-01-01T18:00:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T18:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T19:00:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T19:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T20:00:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T20:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T21:00:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T21:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T22:00:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T22:30:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T23:00:00.000Z,
  },
  Object {
    "fee": 0,
    "hour": 2018-01-01T23:30:00.000Z,
  },
]
`;
